File: cmd/api/graphql_/resolvers/enterprise_resolvers/mutations.go
package enterprise_resolvers

import (
	"go-complaint/application/application_services"
	"go-complaint/application/commands"
	"strings"

	"github.com/graphql-go/graphql"
)

func CreateEnterpriseResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().Credentials(params.Context)
	if err != nil {
		return false, err
	}
	command := commands.EnterpriseCommand{
		OwnerID: currentUser.Email,
		Name:    params.Args["name"].(string),
		Website: params.Args["website"].(string),
		Email:   params.Args["email"].(string),
		//		PhoneCode:      params.Args["phoneCode"].(string),
		Phone:          params.Args["phone"].(string),
		CountryID:      params.Args["countryID"].(int),
		CountryStateID: params.Args["countryStateID"].(int),
		CityID:         params.Args["cityID"].(int),
		IndustryID:     params.Args["industryID"].(int),
		FoundationDate: params.Args["foundationDate"].(string),
	}
	err = command.Register(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func UpdateEnterpriseResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseID"].(string),
		application_services.WRITE,
		"OWNER",
	)
	if err != nil {
		return nil, err
	}
	command := commands.EnterpriseCommand{
		OwnerID:    currentUser.Email,
		Name:       params.Args["enterpriseID"].(string),
		UpdateType: params.Args["updateType"].(string),
	}
	switch command.UpdateType {
	case "logoIMG":
		command.LogoIMG = params.Args["value"].(string)
	case "bannerIMG":
		command.BannerIMG = params.Args["value"].(string)
	case "website":
		command.Website = params.Args["value"].(string)
	case "email":
		command.Email = params.Args["value"].(string)
	case "phone":
		command.Phone = params.Args["value"].(string)
	case "country":
		command.CountryID = params.Args["numberValue"].(int)
	case "countryState":
		command.CountryStateID = params.Args["numberValue"].(int)
	case "city":
		command.CityID = params.Args["numberValue"].(int)
	default:
		return false, command.UpdateEnterprise(params.Context)
	}
	err = command.UpdateEnterprise(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func InviteToEnterpriseResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"OWNER", "MANAGER",
	)
	if err != nil {
		return false, err
	}
	command := commands.EnterpriseCommand{
		OwnerID:   currentUser.Email,
		Name:      params.Args["enterpriseName"].(string),
		ProposeTo: params.Args["proposeTo"].(string),
		Position:  params.Args["proposedPosition"].(string),
	}

	err = command.InviteToProject(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil

}

func HireEmployeeResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"OWNER", "MANAGER",
	)
	if err != nil {
		return false, err
	}
	command := commands.EnterpriseCommand{
		Name:    params.Args["enterpriseName"].(string),
		EventID: params.Args["eventID"].(string),
		OwnerID: currentUser.Email,
	}
	err = command.HireEmployee(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func CancelHiringProccessResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"OWNER", "MANAGER",
	)
	if err != nil {
		return false, err
	}
	reason := ""
	if params.Args["reason"] != nil {
		reason = params.Args["reason"].(string)
	}

	command := commands.EnterpriseCommand{
		OwnerID:      currentUser.Email,
		Name:         params.Args["enterpriseName"].(string),
		EventID:      params.Args["eventID"].(string),
		CancelReason: reason,
	}
	err = command.CancelHiringProccess(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func FireEmployeeResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"OWNER", "MANAGER",
	)
	if err != nil {
		return false, err
	}
	command := commands.EnterpriseCommand{
		OwnerID:    currentUser.Email,
		Name:       params.Args["enterpriseName"].(string),
		EmployeeID: params.Args["employeeID"].(string),
	}
	err = command.FireEmployee(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func PromoteEmployeeResolver(params graphql.ResolveParams) (interface{}, error) {
	currentUser, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"OWNER", "MANAGER",
	)

	if err != nil {
		return false, err
	}
	command := commands.EnterpriseCommand{
		OwnerID:    currentUser.Email,
		Name:       params.Args["enterpriseName"].(string),
		EmployeeID: params.Args["employeeID"].(string),
		Position:   params.Args["position"].(string),
	}
	err = command.PromoteEmployee(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func ChatReplyResolver(params graphql.ResolveParams) (interface{}, error) {
	_, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"ASSISTANT", "OWNER", "MANAGER",
	)
	if err != nil {
		return false, err
	}
	command := commands.EnterpriseCommand{
		ID:       params.Args["id"].(string),
		SenderID: params.Args["senderID"].(string),
		Content:  params.Args["content"].(string),
	}
	err = command.ReplyChat(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

func MarkChatReplyAsSeenResolver(params graphql.ResolveParams) (interface{}, error) {
	_, err := application_services.AuthorizationApplicationServiceInstance().ResourceAccess(
		params.Context,
		"Enterprise",
		params.Args["enterpriseName"].(string),
		application_services.WRITE,
		"ASSISTANT", "OWNER", "MANAGER",
	)
	if err != nil {
		return false, err
	}
	ids := params.Args["repliesID"].(string)
	s := strings.Split(ids, ",")

	command := commands.EnterpriseCommand{
		ID:        params.Args["chatID"].(string),
		RepliesID: s,
	}
	err = command.MarkAsSeen(params.Context)
	if err != nil {
		return false, err
	}
	return true, nil
}

File: presenter/src/lib/mutations.ts
import EmailAlreadyInUseError from "../components/error/EmailAlreadyInUseError";
import { ChangePasswordType } from "../components/profile/settings/settings_lib";
import { csrf } from "./csrf";
import { deleteLinebreaks } from "./delete_line_breaks";
import {
  ContactType,
  CreateAFeedback,
  CreateUser,
  DeclineHiringInvitation,
  EndHiringProcess,
  EnterpriseEventType,
  FireEmployeeType,
  InviteToProject,
  MarkComplaintRepliesAsSeenType,
  MarkReplyChatAsSeenType,
  PromoteEmployeeType,
  RateComplaint,
  RegisterEnterprise,
  ReplyComplaintType,
  ReplyEnterpriseChatType,
  SendComplaintType,
  UpdateEnterpriseType,
  UpdateUserType,
} from "./types";

export const ContactMutation = ({ email, text }: ContactType): string => {
  return `
  mutation {
    Contact(
        email: "${email}",
        text: "${text}"
    )
}
  `;
};

export const CreateUserMutation = ({
  email,
  password,
  firstName,
  lastName,
  gender,
  pronoun,
  birthDate,
  phoneCode,
  phone,
  country,
  county,
  city,
}: CreateUser): string =>
  `
        mutation {
            CreateUser(
                email: "${email}",
                password: "${password}",
                firstName: "${firstName}",
                lastName: "${lastName}",
                gender: "${gender}",
                pronoun: "${pronoun}",
                birthDate: "${birthDate}",
                phoneCode: "${phoneCode}",
                phone: "${phone}",
                country: ${country},
                county: ${county},
                city: ${city},
            )
        }
    `;

export const MarkNotificationAsRead = (id: string): string => `
mutation {
    MarkNotificationAsRead(
        id: "${id}"
    )
}
`;
export const UpdateUserMutation = ({ updateType, value }: UpdateUserType) => {
  return `mutation {
    UpdateUser(
        updateType: "${updateType}",
        value: "${value}"
    )
}
`;
};
export const UpdateUserMutation2 = ({
  updateType,
  numberValue,
}: UpdateUserType) => {
  return `mutation {
    UpdateUser(
        updateType: "${updateType}",
        numberValue: ${numberValue}
    )
}
`;
};
export const ChangePasswordMutation = ({
  oldPassword,
  newPassword,
}: ChangePasswordType): string => `
mutation {
    ChangePassword(
        oldPassword: "${oldPassword}",
        newPassword: "${newPassword}"
    )
}
`;
export const CreateEnterpriseMutation = ({
  name,
  email,
  website,
  phoneCode,
  phone,
  industryID,
  countryID,
  countryStateID,
  cityID,
  foundationDate,
}: RegisterEnterprise): string => `
mutation {
    CreateEnterprise(
        name: "${name}",
        email: "${email}",
        website: "${website}",
        phoneCode: "${phoneCode}",
        phone: "${phone}",
        industryID: ${industryID},
        countryID: ${countryID},
        countryStateID: ${countryStateID},
        cityID: ${cityID},
        foundationDate: "${foundationDate}",
    )
}
`;

export const UpdateEnterpriseMutation = ({
  updateType,
  enterpriseID,
  value,
}: UpdateEnterpriseType) => {
  return `mutation {
    UpdateEnterprise(
        updateType: "${updateType}",
        enterpriseID: "${enterpriseID}",
        value: "${value}"
    )
}
`;
};
export const UpdateEnterpriseMutation2 = ({
  updateType,
  enterpriseID,
  numberValue,
}: UpdateEnterpriseType) => {
  return `mutation {
    UpdateEnterprise(
        updateType: "${updateType}",
        enterpriseID: "${enterpriseID}",
        numberValue: ${numberValue}
    )
}
`;
};

export const HireEmployeeMutation = ({
  enterpriseName,
  eventID,
}: EnterpriseEventType) => {
  return `
  mutation {
    HireEmployee(
      enterpriseName: "${enterpriseName}",
      eventID: "${eventID}",
    )
  }
  `;
};
export const CancelHiringProccessMutation = ({
  enterpriseName,
  eventID,
}: EnterpriseEventType) => {
  return `
  mutation {
    CancelHiringProccess(
      enterpriseName: "${enterpriseName}",
      eventID: "${eventID}",
    )
  }
  `;
};
export const SendComplaintMutation = ({
  authorID,
  receiverID,
  receiverFullName,
  receiverProfileIMG,
  title,
  description,
  content,
}: SendComplaintType): string => `
mutation {
    SendComplaint(
        authorID: "${authorID}",
        receiverID: "${receiverID}",
        receiverFullName: "${receiverFullName}",
        receiverProfileIMG: "${receiverProfileIMG}",
        title: "${title}",
        description: "${description}",
        content: "${content}",
    )
}
`;

export const ReplyComplaintMutation = ({
  complaintID,
  replyAuthorID,
  replyBody,
  replyEnterpriseID,
}: ReplyComplaintType): string => `
mutation {
    ReplyComplaint(
        complaintID: "${complaintID}",
        replyAuthorID: "${replyAuthorID}",
        replyBody: "${replyBody}",
        replyEnterpriseID: "${replyEnterpriseID}",
    )
}
`;

export const MarkComplaintRepliesAsSeenMutation = ({
  complaintID,
  repliesID,
}: MarkComplaintRepliesAsSeenType): string => {
  return `
mutation {
    MarkAsSeen(
        complaintID: "${complaintID}",
        ids: "${repliesID.join(",")}",
    )
}
`;
};

export const InviteToProjectMutation = ({
  enterpriseName,
  proposedPosition,
  proposeTo,
}: InviteToProject): string => {
  return `
  mutation {
    InviteToEnterprise(
      enterpriseName: "${enterpriseName}",
      proposedPosition: "${proposedPosition}",
      proposeTo: "${proposeTo}"
    )
  }
  `;
};

export const AcceptHiringInvitationMutation = (id: string): string => {
  return `
  mutation {
    AcceptEnterpriseInvitation(
      id: "${id}"
    )
  }
  `;
};
export const DeclineHiringInvitationMutation = ({
  id,
  reason,
}: DeclineHiringInvitation): string => {
  return `
  mutation {
    RejectEnterpriseInvitation(
      id: "${id}"
      reason: "${reason}"
    )
  }
  `;
};
export const EndHiringProcessMutation = (ehp: EndHiringProcess): string => {
  return `mutation {
    EndHiringProcess(
      pendingEventID: "${ehp.pendingEventID}",
      enterpriseName: "${ehp.enterpriseID}",
      accepted: ${ehp.accepted}
    )
  }
  `;
};
export type LeaveEnterpriseType = {
  enterpriseName: string;
  employeeID: string;
};
export const LeaveEnterpriseMutation = ({
  enterpriseName,
  employeeID,
}: LeaveEnterpriseType): string => {
  return `mutation {
    LeaveEnterprise(
      enterpriseName: "${enterpriseName}",
      employeeID: "${employeeID}"
    )
  }
  `;
};

export const MarkAsReviewableMutation = (id: string): string => {
  return `mutation {
    SendForReviewing(
      id: "${id}",
    )
  }
  `;
};

export const RateComplaintMutation = (rate: RateComplaint): string => {
  return `mutation {
    RateComplaint(
      complaintId: "${rate.complaintId}",
      eventId: "${rate.eventId}",
      rate: ${rate.rate},
      comment: "${rate.comment}",
      )
    }`;
};

export const CreateFeedbackMutation = ({
  enterpriseID,
  complaintID,
  color,
}: CreateAFeedback) => {
  return `
  mutation{
    CreateFeedback(
      enterpriseID: "${enterpriseID}",
      complaintID: "${complaintID}",
      color: "${color}",
      )
    }`;
};
export const AddReplyMutation = ({
  enterpriseID,
  feedbackID,
  reviewerID,
  color,
  repliesID,
}: CreateAFeedback) => {
  return `
  mutation{
    AddReply(
      enterpriseID: "${enterpriseID}",
      feedbackID: "${feedbackID}",
      reviewerID: "${reviewerID}",
      color: "${color}",
      repliesID: "${repliesID}",
      )
      }`;
};

export const RemoveReplyMutation = ({
  enterpriseID,
  feedbackID,
  color,
  repliesID,
}: CreateAFeedback) => {
  return `
  mutation{
    RemoveReply(
      enterpriseID: "${enterpriseID}",
      feedbackID: "${feedbackID}",
      color: "${color}",
      repliesID: "${repliesID}",
      )
      }`;
};

export const AddCommentMutation = ({
  enterpriseID,
  feedbackID,
  color,
  comment,
}: CreateAFeedback) => {
  return `
  mutation{
    AddComment(
      enterpriseID: "${enterpriseID}",
      feedbackID: "${feedbackID}",
      color: "${color}",
      comment: "${comment}",
      )
      }`;
};

export const DeleteCommentMutation = ({
  enterpriseID,
  feedbackID,
  color,
}: CreateAFeedback) => {
  return `
  mutation{
    DeleteComment(
      enterpriseID: "${enterpriseID}",
      feedbackID: "${feedbackID}",
      color: "${color}",
      )
      }`;
};

export const EndFeedbackMutation = ({
  enterpriseID,
  feedbackID,
}: CreateAFeedback) => {
  return `
  mutation{
    EndFeedback(
      enterpriseID: "${enterpriseID}",
      feedbackID: "${feedbackID}",
      )
    }`;
};

export const ReplyChatMutation = ({
  id,
  enterpriseName,
  senderID,
  content,
}: ReplyEnterpriseChatType): string => `
mutation {
    ReplyChat(
        id: "${id}",
        enterpriseName: "${enterpriseName}",
        senderID: "${senderID}",
        content: "${content}",
    )
}
`;

export const MarkReplyChatAsSeenMutation = ({
  chatID,
  enterpriseName,
  repliesID,
}: MarkReplyChatAsSeenType): string => {
  return `
mutation {
    MarkReplyChatAsSeen(
        chatID: "${chatID}",
        enterpriseName: "${enterpriseName}",
        repliesID: "${repliesID.join(",")}",
    )
}
`;
};

export const PromoteEmployeeMutation = ({
  enterpriseName,
  employeeID,
  position,
}: PromoteEmployeeType): string => `
mutation{
  PromoteEmployee(
    enterpriseName:"${enterpriseName}",
    employeeID:"${employeeID}",
    position:"${position}"
  )
}
`;

export const FireEmployeeMutation = ({
  enterpriseName,
  employeeID,
}: FireEmployeeType): string => `
mutation{
  FireEmployee(
    enterpriseName:"${enterpriseName}",
    employeeID:"${employeeID}"
  )
}
`;

export const Mutation = async <T>(
  mutationFn: (data: T) => string,
  arg: T
): Promise<boolean> => {
  const token = await csrf();
  if (token != "") {
    const response = await fetch("https://api.go-complaint.com/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-csrf-token": token,
      },
      credentials: "include",
      body: JSON.stringify({ query: deleteLinebreaks(mutationFn(arg)) }),
    });
    const data = await response.json();
    if (data.errors) {
      const stringErr = data.errors[0].message;
      if (stringErr.includes("SQLSTATE 23505")) {
        throw new EmailAlreadyInUseError();
      }
      throw new Error(stringErr);
    }
    return true;
  }
  throw new Error("No CSRF token");
};
